leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/QueueModeLwM2mRequestSender.java

```diff
@@ -0,0 +1,164 @@
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *     RISE SICS AB - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.leshan.server.californium.impl;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.leshan.core.request.DownlinkRequest;
+import org.eclipse.leshan.core.request.exception.TimeoutException;
+import org.eclipse.leshan.core.response.ErrorCallback;
+import org.eclipse.leshan.core.response.LwM2mResponse;
+import org.eclipse.leshan.core.response.ResponseCallback;
+import org.eclipse.leshan.server.queue.Presence;
+import org.eclipse.leshan.server.queue.PresenceServiceImpl;
+import org.eclipse.leshan.server.registration.Registration;
+import org.eclipse.leshan.server.request.LwM2mRequestSender;
+import org.eclipse.leshan.util.Validate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class QueueModeLwM2mRequestSender implements LwM2mRequestSender {
+
+    private static final Logger LOG = LoggerFactory.getLogger(QueueModeLwM2mRequestSender.class);
+
+    private PresenceServiceImpl presenceService;
+    private LwM2mRequestSender delegatedSender;
+
+    /**
+     * @param presenceService the presence service object for setting the client into {@link Presence#SLEEPING} when
+     *        request Timeout expires and into {@link Presence#Awake} when a response arrives.
+     * @param delegatedSender internal sender that it is used for sending the requests, using delegation.
+     */
+    public QueueModeLwM2mRequestSender(PresenceServiceImpl presenceService, LwM2mRequestSender delegatedSender) {
+        Validate.notNull(presenceService);
+        Validate.notNull(delegatedSender);
+
+        this.presenceService = presenceService;
+        this.delegatedSender = delegatedSender;
+    }
+
+    @Override
+    public <T extends LwM2mResponse> T send(final Registration destination, DownlinkRequest<T> request, long timeout)
+            throws InterruptedException {
+
+        // If the client is sleeping, warn the user and return
+        if (!presenceService.isClientAwake(destination)) {
+            LOG.info("The destination client is sleeping, request couldn't been sent.");
+            return null;
+        }
+
+        // Create a response callback to use the asynchronous send method in a synchronous way
+        ResponseCallbackQueueModeSynchronous<T> synchronousResponseCallback = new ResponseCallbackQueueModeSynchronous<T>(
+                presenceService, destination);
```

Comment:
Why we don't use the  delegatedSender synchronous API ?
(I remember you already try to explain me that but I probably miss-understood)

Reply:
We cannot directly use the function `public <T extends LwM2mResponse> T send(final Registration destination, final DownlinkRequest<T> request, long timeout)` because it is inside this function where the message observer is added to the coap request. Then, we cannot change it to perform the required actions in the methods onResponse() and onTimeout() (or at least I don't see a way).

So this method could be change to take the SyncRequestObserver as a parameter, and create a new one that overrides the onResponse() and onTimeout() methods. Or on the other hand we can use the asynchronous method with the callback that I created to make it synchronous.

Reply:
Why not something like this ? 
```java
T response = delegatedSender.send(destination, request, timeout);
        if (response != null) {
            presenceService.setAwake(destination);
        } else {
            presenceService.clientNotResponding(destination);
        }
        return response;
```

Reply:
You're totally right... the `waitForResponse()` will return when a response or a timeout occurs. 

Sometimes we try to do things much more complicated than they really are :-).

I'll change this two things from your comments and come with a new commit!